File: intel/src/main/java/au/com/codeka/warworlds/intel/RequestException.java
package au.com.codeka.warworlds.intel;

import javax.servlet.http.HttpServletResponse;

/**
 * This exception is thrown when you want to pass an error back to the client.
 */
public class RequestException extends Exception {
  private static final long serialVersionUID = 1L;

  private int httpErrorCode;

  public RequestException(int httpErrorCode) {
    super(String.format("HTTP Error: %d", httpErrorCode));
    this.httpErrorCode = httpErrorCode;
  }

  public RequestException(int httpErrorCode, String message) {
    super(String.format(message, httpErrorCode));
    this.httpErrorCode = httpErrorCode;
  }

  public RequestException(int httpErrorCode, String message, Throwable innerException) {
    super(String.format(message, httpErrorCode), innerException);
    this.httpErrorCode = httpErrorCode;
  }

  public void populate(HttpServletResponse response) {
    response.setStatus(httpErrorCode);
  }

  public int getHttpErrorCode() {
    return httpErrorCode;
  }
}

File: server/src/main/kotlin/au/com/codeka/warworlds/server/handlers/RequestException.kt
package au.com.codeka.warworlds.server.handlers

import java.util.*
import javax.servlet.http.HttpServletResponse

/** This exception is thrown when you want to pass an error back to the client. */
class RequestException : Exception {
  var errorCode = 0
    private set

  constructor(httpErrorCode: Int)
      : super(String.format(Locale.US, "HTTP Error: %d", httpErrorCode)) {
    errorCode = httpErrorCode
  }

  constructor(httpErrorCode: Int, message: String?)
      : super(String.format(message!!, httpErrorCode)) {
    errorCode = httpErrorCode
  }

  constructor(httpErrorCode: Int, message: String?, innerException: Throwable?)
      : super(String.format(message!!, httpErrorCode), innerException) {
    errorCode = httpErrorCode
  }

  constructor(innerException: Throwable)
      : super(getExceptionDescription(innerException), innerException) {
    val reqExc = findInnerException(innerException, RequestException::class.java)
    errorCode = reqExc?.errorCode ?: 500
  }

  fun populate(response: HttpServletResponse?) {
    response!!.status = errorCode
  }

  companion object {
    private const val serialVersionUID = 1L
    private fun getExceptionDescription(e: Throwable): String {
      val reqExc = findInnerException(e, RequestException::class.java)
      return if (reqExc != null) {
        "HTTP Error: " + reqExc.errorCode
      } else "Unknown Exception"
    }

    private fun <T : Exception?> findInnerException(e: Throwable, exceptionType: Class<T>): T? {
      var inner: Throwable? = e
      while (inner != null) {
        if (inner.javaClass == exceptionType) {
          @Suppress("UNCHECKED_CAST")
          return inner as T
        }
        inner = inner.cause
      }
      return null
    }
  }
}